### ISSUE TITLE
Fighting deadlocks

### ISSUE THREAD

**Issue by zixeljupo** [pytorch/fairseq#3509]
## ❓ Questions and Help

I wrote a lot of code on the top of Fairseq repository (mostly new models, new evaluation script, some changes for model training in Trainer class), but it should not affect multiprocessing/multithreading or related stuff. Each launch after about 1.5 to 2.5 epochs (in 7-GPU setup) I am experiencing deadlock. During this deadlock, all 7 GPUs are at 100% load.

This deadlock was looking like a problem in the wrong calls to torch.distributed, but after I patched most torch.distributed methods to log their calls to separate files, I found that this is not the case (but I might forget about some methods).

The question: what should I use to find the actual line of code which causes deadlock? As I understand, Fairseq uses spawn method to create GPU workers, and calling pdb, for example, will not help (It will show ^C for main thread), not for workers.

**Comment by villmow** [comment]
Had the same symptoms half a year ago, once one GPU received an OOM the others started having a 100% load.

**Comment by zixeljupo** [comment]
> Had the same symptoms half a year ago, once one GPU received an OOM the others started having a 100% load.

Usually, when one worker OOMs or dies with another exception, this frees GPU for this worker, while others hang on distributed communication, which shows 100% load for other GPUs. However, in my case, all 7 (out of 7 available) GPUs are at 100% load. Did you have 100% load at all selected GPUs

**Comment by zixeljupo** [comment]
Ok, this was a bug on my side. The problem was with custom gradient synchronization and with the fact that sometimes some of the gradients were None.
